File: backend/src/domain/gate/security-overview.ts
import { GateIssue, GateVerdict, evaluate } from './policy-gate';
import { PolicyLookup, ResolvedPolicy, StoredPolicy, resolvePolicy } from './policy-resolution';

/**
 * Where each target stands, in one picture.
 *
 * **The gate verdict was always computed and never shown.** It served
 * `POST /api/v1/gate`, and a team could only learn whether its repository passed by
 * running a build against it — the application knew the answer and kept it to itself.
 *
 * ## Two rules shape this implementation
 *
 * **The verdict here must be the one the API returns.** This module therefore computes no
 * pass/fail of its own: it calls `evaluate` with the same policy resolution as the
 * endpoint, over the same issues. A SQL aggregate recounting "open issues above the
 * threshold" would agree today and diverge the first time a flag was added to
 * `GatePolicy` — and nobody would see it until a pipeline and a screen contradicted each
 * other about the same repository.
 *
 * **A screen listing N targets must not cost N queries.** Both traps are real: resolving
 * a policy per target costs one or two queries each, loading a target's issues costs
 * another. Everything is therefore read once and matched up in memory here — hence a pure
 * function over already-loaded data, rather than a service holding a session.
 *
 * **A target never scanned, or whose last scan failed, is not a target that passes.** It
 * is a target nobody has looked at — the worst posture there is, and the one no screen
 * named. An empty backlog passes every policy: saying so without that qualifier would be
 * the most misleading thing this screen could do.
 */

export const TARGET_REPOSITORY = 'repository';
export const TARGET_CONTAINER = 'container';

/** What the last scan says about how much the verdict can be trusted. */
export const OBSERVATION_OK = 'ok';
export const OBSERVATION_NEVER_SCANNED = 'never_scanned';
export const OBSERVATION_LAST_SCAN_FAILED = 'last_scan_failed';
export const OBSERVATION_IN_PROGRESS = 'in_progress';

export type Observation = typeof OBSERVATION_OK | typeof OBSERVATION_NEVER_SCANNED | typeof OBSERVATION_LAST_SCAN_FAILED | typeof OBSERVATION_IN_PROGRESS;

const IN_FLIGHT_STATUSES = ['pending', 'scanning'];

export interface OverviewTarget {
    id: number;
    /** The readable name; for a repository, its name or failing that its URL. */
    name: string;
}

export interface LatestScan {
    id: number;
    status: string | null;
    createdAt: Date | null;
}

export interface TargetPosture {
    kind: typeof TARGET_REPOSITORY | typeof TARGET_CONTAINER;
    targetId: number;
    name: string;
    verdict: GateVerdict;
    policy: ResolvedPolicy;
    observation: Observation;
    lastScanAt: Date | null;
    lastScanId: number | null;
    passed: boolean;
    /**
     * Does the verdict rest on a real observation?
     *
     * A target nobody has successfully scanned produces an empty backlog, and an empty
     * backlog passes every policy.
     */
    observed: boolean;
}

export interface SecurityOverview {
    targets: TargetPosture[];
    failingCount: number;
    totalCount: number;
    kevCount: number;
    neverScannedCount: number;
    lastScanFailedCount: number;
}

export interface OverviewInput {
    repositories: OverviewTarget[];
    containers: OverviewTarget[];
    /** Every active policy, read in one go. */
    policies: { targetKind: string; targetId: number; policy: StoredPolicy }[];
    /** Every open issue, read in one go. */
    openIssues: (GateIssue & { repoId: number | null; containerId: number | null })[];
    latestScanByRepository: Map<number, LatestScan>;
    latestScanByContainer: Map<number, LatestScan>;
}

/** Assembles the view from already-read data. No queries here, by construction. */
export function buildOverview(input: OverviewInput): SecurityOverview {
    const byScope = new Map<string, StoredPolicy>();
    for (const entry of input.policies) byScope.set(scopeKey(entry.targetKind, entry.targetId), entry.policy);

    const issuesByTarget = new Map<string, GateIssue[]>();
    for (const issue of input.openIssues) {
        const key = issue.repoId != null ? scopeKey(TARGET_REPOSITORY, issue.repoId) : scopeKey(TARGET_CONTAINER, issue.containerId ?? -1);
        const bucket = issuesByTarget.get(key);
        if (bucket) bucket.push(issue);
        else issuesByTarget.set(key, [issue]);
    }

    const targets: TargetPosture[] = [
        ...input.repositories.map((repository) => posture(TARGET_REPOSITORY, repository, byScope, issuesByTarget, input.latestScanByRepository.get(repository.id))),
        ...input.containers.map((container) => posture(TARGET_CONTAINER, container, byScope, issuesByTarget, input.latestScanByContainer.get(container.id)))
    ];

    return {
        targets,
        failingCount: targets.filter((target) => !target.passed).length,
        totalCount: targets.length,
        // Counted over the evaluated issues, not over the whole backlog: a KEV discarded
        // by a triage or by `fixableOnly` does not weigh on the verdict, and showing it in
        // the same banner would present a number that corresponds to nothing.
        kevCount: targets.reduce((total, target) => total + target.verdict.violations.filter((violation) => violation.rule === 'kev').length, 0),
        neverScannedCount: targets.filter((target) => target.observation === OBSERVATION_NEVER_SCANNED).length,
        lastScanFailedCount: targets.filter((target) => target.observation === OBSERVATION_LAST_SCAN_FAILED).length
    };
}

function posture(
    kind: typeof TARGET_REPOSITORY | typeof TARGET_CONTAINER,
    target: OverviewTarget,
    byScope: Map<string, StoredPolicy>,
    issuesByTarget: Map<string, GateIssue[]>,
    latestScan: LatestScan | undefined
): TargetPosture {
    // The same precedence as `resolvePolicy`, applied to policies read only once: calling
    // it per target is exactly what would make this screen 2N queries.
    const lookup: PolicyLookup = {
        forTarget: byScope.get(scopeKey(kind, target.id)) ?? null,
        global: byScope.get(GLOBAL_KEY) ?? null
    };
    const policy = resolvePolicy(lookup);
    const observation = observationOf(latestScan);

    const verdict = evaluate(issuesByTarget.get(scopeKey(kind, target.id)) ?? [], policy.policy);

    return {
        kind,
        targetId: target.id,
        name: target.name,
        verdict,
        policy,
        observation,
        lastScanAt: latestScan?.createdAt ?? null,
        lastScanId: latestScan?.id ?? null,
        passed: verdict.passed,
        observed: observation === OBSERVATION_OK
    };
}

function observationOf(latestScan: LatestScan | undefined): Observation {
    if (!latestScan) return OBSERVATION_NEVER_SCANNED;
    if (latestScan.status && IN_FLIGHT_STATUSES.includes(latestScan.status)) return OBSERVATION_IN_PROGRESS;
    if (latestScan.status === 'failed') return OBSERVATION_LAST_SCAN_FAILED;
    return OBSERVATION_OK;
}

/** The global policy is stored with the `global` scope and identifier 0. */
const GLOBAL_KEY = 'global:0';

function scopeKey(kind: string, id: number): string {
    return `${kind}:${id}`;
}

File: vectispire-java/vectispire-common/src/main/java/com/asmolabs/vectispire/common/domain/gate/SecurityOverview.java
package com.asmolabs.zanshin.common.domain.gate;

import com.asmolabs.zanshin.common.domain.scans.ScanStatus;
import com.asmolabs.zanshin.common.domain.targets.ScanTarget;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;

/**
 * Where each target stands, in one picture.
 *
 * <p><b>The gate verdict was always computed and never shown.</b> It served the gate endpoint,
 * and a team could only learn whether its repository passed by running a build against it —
 * the application knew the answer and kept it to itself.
 *
 * <h2>Two rules shape this implementation</h2>
 *
 * <p><b>The verdict here must be the one the API returns.</b> This class computes no pass/fail
 * of its own: it calls {@link PolicyGate#evaluate} with the same resolution the endpoint uses,
 * over the same issues. A SQL aggregate recounting "open issues above the threshold" would
 * agree today and diverge the first time a flag was added to {@link GatePolicy} — and nobody
 * would notice until a pipeline and a screen contradicted each other about one repository.
 *
 * <p><b>A screen listing N targets must not cost N queries.</b> Both traps are real: resolving
 * a policy per target costs one or two queries each, loading a target's issues costs another.
 * Everything is read once and matched up here — hence a pure function over already-loaded
 * data, rather than a service holding a session.
 *
 * <h2>An empty backlog is not a good posture</h2>
 *
 * <p><b>A target never scanned, or whose last scan failed, is not a target that passes.</b> It
 * is a target nobody has looked at — the worst posture there is, and the one no screen named.
 * An empty backlog passes every policy; saying so without the qualifier would be the most
 * misleading thing this screen could do. That is what {@link TargetPosture#observed()} carries,
 * and why it is a separate field from {@code passed}.
 */
public final class SecurityOverview {

    private SecurityOverview() {}

    /** What the last scan says about how much the verdict can be trusted. */
    public enum Observation {
        OK,
        NEVER_SCANNED,
        LAST_SCAN_FAILED,
        IN_PROGRESS
    }

    /** A target as this screen names it. */
    public record NamedTarget(ScanTarget target, String name) {}

    public record LatestScan(long id, ScanStatus status, Instant createdAt) {}

    /**
     * @param observed whether the verdict rests on a real observation — a target nobody has
     *     successfully scanned produces an empty backlog, and an empty backlog passes
     *     everything
     */
    public record TargetPosture(
            ScanTarget target,
            String name,
            GateVerdict verdict,
            PolicyResolution.ResolvedPolicy policy,
            Observation observation,
            Optional<LatestScan> lastScan,
            boolean passed,
            boolean observed) {}

    public record Overview(
            List<TargetPosture> targets,
            int failingCount,
            int totalCount,
            long kevCount,
            long neverScannedCount,
            long lastScanFailedCount) {

        public Overview {
            targets = List.copyOf(targets);
        }
    }

    /**
     * Everything the screen needs, already read.
     *
     * <p>The global policy is a field of its own rather than a row keyed {@code "global:0"} in
     * the same map as the targets. The original used that magic key, which meant a repository
     * whose scope key ever collided with it would silently inherit the wrong rules.
     *
     * @param openIssues every open issue, read in one go, grouped here rather than per target
     */
    public record Input(
            List<NamedTarget> targets,
            Map<ScanTarget, PolicyResolution.StoredPolicy> policiesByTarget,
            Optional<PolicyResolution.StoredPolicy> globalPolicy,
            Map<ScanTarget, List<GateIssue>> openIssues,
            Map<ScanTarget, LatestScan> latestScans) {}

    /** Assembles the view from already-read data. No queries here, by construction. */
    public static Overview build(Input input) {
        List<TargetPosture> postures = new ArrayList<>(input.targets().size());
        for (NamedTarget named : input.targets()) {
            postures.add(posture(named, input));
        }

        return new Overview(
                postures,
                (int) postures.stream().filter(posture -> !posture.passed()).count(),
                postures.size(),
                // Counted over the *evaluated* issues, not the whole backlog: a KEV discarded by
                // a triage decision or by `fixableOnly` does not weigh on the verdict, and
                // showing it in the same banner would present a number corresponding to nothing.
                postures.stream()
                        .flatMap(posture -> posture.verdict().violations().stream())
                        .filter(violation -> violation.rule() == GateVerdict.Rule.KEV)
                        .count(),
                postures.stream().filter(p -> p.observation() == Observation.NEVER_SCANNED).count(),
                postures.stream().filter(p -> p.observation() == Observation.LAST_SCAN_FAILED).count());
    }

    private static TargetPosture posture(NamedTarget named, Input input) {
        // The same precedence as `PolicyResolution.resolve`, over policies read once. Calling
        // the repository per target is exactly what would make this screen cost 2N queries.
        PolicyResolution.PolicyLookup lookup = new PolicyResolution.PolicyLookup(
                Optional.ofNullable(input.policiesByTarget().get(named.target())), input.globalPolicy());
        PolicyResolution.ResolvedPolicy policy =
                PolicyResolution.resolve(lookup, RequestedPolicy.none(), PolicyResolution.Scope.TARGET);

        Optional<LatestScan> latest = Optional.ofNullable(input.latestScans().get(named.target()));
        Observation observation = observationOf(latest);

        GateVerdict verdict = PolicyGate.evaluate(
                input.openIssues().getOrDefault(named.target(), List.of()), policy.policy());

        return new TargetPosture(
                named.target(),
                named.name(),
                verdict,
                policy,
                observation,
                latest,
                verdict.passed(),
                observation == Observation.OK);
    }

    private static Observation observationOf(Optional<LatestScan> latest) {
        if (latest.isEmpty()) {
            return Observation.NEVER_SCANNED;
        }
        ScanStatus status = latest.get().status();
        if (status == null) {
            return Observation.OK;
        }
        if (status.isInFlight()) {
            return Observation.IN_PROGRESS;
        }
        return status == ScanStatus.FAILED ? Observation.LAST_SCAN_FAILED : Observation.OK;
    }
}
